repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -0,0 +1,150 @@
+import scala.collection.immutable.{TreeMap, TreeSet}
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: adrian
+ * Date: 5/27/15
+ * Time: 10:25 AM
+ */
+class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream: ServletInputStream)
+  extends javax.servlet.http.HttpServletRequestWrapper(originalRequest)
+  with HeaderInteractor {
+
+  object RequestBodyStatus extends Enumeration {
+    val Available, InputStream, Reader = Value
+  }
+
+  private var status = RequestBodyStatus.Available
+
+  def this(originalRequest: HttpServletRequest) = this(originalRequest, originalRequest.getInputStream())
+
+  val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+
+  private var headerMap: Map[String, List[String]] = new TreeMap[String, List[String]]()(caseInsensitiveOrdering)
+  private var removedHeaders: Set[String] = new TreeSet[String]()(caseInsensitiveOrdering)
+
+  override def getInputStream: ServletInputStream = {
+    if (status == RequestBodyStatus.Reader) throw new IllegalStateException else status = RequestBodyStatus.InputStream
+    inputStream
+  }
+
+  override def getReader: BufferedReader = {
+    if (status == RequestBodyStatus.InputStream) throw new IllegalStateException else status = RequestBodyStatus.Reader
+    new BufferedReader(new InputStreamReader(inputStream))
+  }
+
+  def getHeaderNamesScala: Set[String] = headerMap.keySet ++ super.getHeaderNames.asScala.toSet.filterNot(removedHeaders.contains)
+
+  override def getHeaderNames: util.Enumeration[String] = getHeaderNamesScala.toIterator.asJavaEnumeration
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNamesScala.toList.asJava
+
+  override def getIntHeader(headerName: String): Int = Option(getHeader(headerName)).getOrElse("-1").toInt
+
+  def getHeadersScala(headerName: String): List[String] = {
+    if (removedHeaders.contains(headerName)) {
+      List[String]()
+    }
+    else {
+      headerMap.getOrElse(headerName, super.getHeaders(headerName).asScala.toList)
+    }
+  }
+
+  override def getHeaders(headerName: String): util.Enumeration[String] = getHeadersScala(headerName).toIterator.asJavaEnumeration
+
+  override def getDateHeader(headerName: String): Long = Option(getHeader(headerName)).map(DateUtils.parseDate(_).getTime).getOrElse(-1)
+
+  override def getHeader(headerName: String): String = getHeadersScala(headerName).headOption.orNull
+
+  override def getHeadersList(headerName: String): util.List[String] = getHeadersScala(headerName).asJava
+
+  override def addHeader(headerName: String, headerValue: String): Unit = {
+    val existingHeaders: List[String] = getHeadersScala(headerName) //this has to be done before we remove from the list,
+                                                                    // because getting this list is partially based on the contents of the removed list
+    if (removedHeaders.contains(headerName)) {
+      removedHeaders = removedHeaders - headerName
+    }
```

Comment:
Thinking about it, this conditional is unnecessary. We can always attempt to remove headerName, although it will only ever have an effect if headerName is in the set.

That is, we can simplify the three lines here to just the middle line.